cpp/src/fil/internal.cuh

```diff
@@ -75,49 +79,71 @@ enum output_t {
 union val_t {
   /** threshold value for branch node or output value (e.g. class
       probability or regression summand) for leaf node */
-  float f;
-  /** class label */
+  float f = NAN;
+  /** class label, leaf vector index or categorical node set offset */
   int idx;
 };
 
 /** base_node contains common implementation details for dense and sparse nodes */
 struct base_node {
-  /** val is either the threshold (for inner nodes, always float)
-      or the tree prediction (for leaf nodes) */
+  /** val, for branch nodes, is a threshold or category list offset. For leaf
+      nodes, it is the tree prediction (see see leaf_output_t<leaf_algo_t>::T) */
   val_t val;
   /** bits encode various information about the node, with the exact nature of
       this information depending on the node type; it includes e.g. whether the
       node is a leaf or inner node, and for inner nodes, additional information,
       e.g. the default direction, feature id or child index */
   int bits;
-  static const int FID_MASK      = (1 << 30) - 1;
-  static const int DEF_LEFT_MASK = 1 << 30;
-  static const int IS_LEAF_MASK  = 1 << 31;
+  static const int IS_LEAF_OFFSET        = 31;
+  static const int IS_LEAF_MASK          = 1 << IS_LEAF_OFFSET;
+  static const int DEF_LEFT_OFFSET       = IS_LEAF_OFFSET - 1;
+  static const int DEF_LEFT_MASK         = 1 << DEF_LEFT_OFFSET;
+  static const int IS_CATEGORICAL_OFFSET = DEF_LEFT_OFFSET - 1;
+  static const int IS_CATEGORICAL_MASK   = 1 << IS_CATEGORICAL_OFFSET;
+  static const int FID_MASK              = (1 << IS_CATEGORICAL_OFFSET) - 1;
   template <class o_t>
-  __host__ __device__ o_t output() const
-  {
-    return val;
-  }
+  __host__ __device__ o_t output() const;
+  __host__ __device__ int set() const { return val.idx; }
   __host__ __device__ float thresh() const { return val.f; }
+  __host__ __device__ val_t split() const { return val; }
```

Comment:
Alternative is something like `{.idx = node.set()}` in the tests, which is confusing (we rely on internal structure of `val_t` that all valid floats will pass fine as ints).

Or, rename `output()` to something generic, and rely on internal structure of `base_node` that output and split are stored in the same place. Could maybe come up with a 4-way union instead of all these?